source/s3_auto_ranged_put.c

```diff
@@ -800,23 +887,90 @@ static int s_s3_auto_ranged_put_prepare_request(
             size_t request_body_size = s_compute_request_body_size(meta_request, request->part_number);
 
             if (request->num_times_prepared == 0) {
-                if (s_skip_parts_from_stream(
-                        meta_request,
-                        auto_ranged_put->prepare_data.num_parts_read_from_stream,
-                        request->part_number - 1)) {
-                    goto message_create_failed;
+                if (auto_ranged_put->has_content_length) {
+                    if (s_skip_parts_from_stream(
+                            meta_request,
+                            auto_ranged_put->prepare_data.num_parts_read_from_stream,
+                            request->part_number - 1)) {
+                        goto message_create_failed;
+                    }
                 }
-                auto_ranged_put->prepare_data.num_parts_read_from_stream = request->part_number - 1;
 
-                aws_byte_buf_init(&request->request_body, meta_request->allocator, request_body_size);
+                /* Some streams might require additional read to realize that
+                 * there is no more data to read. So lets try to read and mark
+                 * part as no_op is we cannot read anymore. */
+                bool successfully_read_data_from_stream = false;
+                if (!aws_s3_meta_request_body_has_no_more_data(meta_request)) {
+                    aws_byte_buf_init(&request->request_body, meta_request->allocator, request_body_size);
+                    if (aws_s3_meta_request_read_body(meta_request, &request->request_body)) {
+                        goto message_create_failed;
+                    }
 
-                if (aws_s3_meta_request_read_body(meta_request, &request->request_body)) {
-                    goto message_create_failed;
+                    successfully_read_data_from_stream = request->request_body.len != 0;
```

Comment:
what if the input stream is not end, but just stalled (no data available at the moment, but probably will have more data later)?

Reply:
discussed in one of the comments above. 
thats the current interface contract we have, user needs to always fill in the buffer we give them even if they dont have data readily available, forcing them to block, etc... 
I think we all agree that its a bad contract, but changing that contract is not part of this PR. Async streaming will resolve ehta

Reply:
aws-c-s3 already suffers from that bug (the current code assumes that every attempt to read will completely fill the buffer)

it will be fixed when I do async-streaming

Reply:
Hmmm, but we are not error out or anything to enforce the assumption, it will continue and probably upload the file with skipping part number with current impl, right?

Reply:
Oh, if this is an existing bug, then, yeah, sure. :live-with-it

Reply:
yeah, we currently dont enforce that assumption, which is a bit scary. we can probably put in a fatal if it has not read the whole body